<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.335">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="description" content="The first session about the notebook is about exploring https://www.malariagen.net/.">

<title>workshop-kenya - 1. Session - MalariaGEN</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script>
<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>


<link rel="stylesheet" href="styles.css">
<meta property="og:title" content="workshop-kenya - 1. Session - MalariaGEN">
<meta property="og:description" content="The first session about the notebook is about exploring `https://www.malariagen.net/`.">
<meta property="og:site-name" content="workshop-kenya">
<meta name="twitter:title" content="workshop-kenya - 1. Session - MalariaGEN">
<meta name="twitter:description" content="The first session about the notebook is about exploring `https://www.malariagen.net/`.">
<meta name="twitter:card" content="summary">
</head>

<body class="nav-sidebar floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">workshop-kenya</span>
    </a>
  </div>
          <div id="quarto-search" class="" title="Search"></div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
    <div class="container-fluid d-flex justify-content-between">
      <h1 class="quarto-secondary-nav-title">1. Session - MalariaGEN</h1>
      <button type="button" class="quarto-btn-toggle btn" aria-label="Show secondary navigation">
        <i class="bi bi-chevron-right"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse sidebar-navigation floating overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">Summer School Epidemiology 2023</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./introduction_to_deep_learning.html" class="sidebar-item-text sidebar-link">Introduction to Deep Learning (DL)</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./installation-guide.html" class="sidebar-item-text sidebar-link">Installation guide</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session_1.html" class="sidebar-item-text sidebar-link active">1. Session - MalariaGEN</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session_1_example.html" class="sidebar-item-text sidebar-link">1. Session (shortened)</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session_2.html" class="sidebar-item-text sidebar-link">2. Session - Introduction to Deep Learning and Dataset Generation</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session_3.html" class="sidebar-item-text sidebar-link">3. Session - Training and Inference of Coalescent Models</a>
  </div>
</li>
    </ul>
    </div>
</nav>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#data" id="toc-data" class="nav-link active" data-scroll-target="#data">Data</a></li>
  <li><a href="#outline" id="toc-outline" class="nav-link" data-scroll-target="#outline">Outline</a></li>
  <li><a href="#analysis" id="toc-analysis" class="nav-link" data-scroll-target="#analysis">Analysis</a>
  <ul class="collapse">
  <li><a href="#metadata" id="toc-metadata" class="nav-link" data-scroll-target="#metadata">Metadata</a></li>
  </ul></li>
  <li><a href="#what-is-an-gff-file-general-feature-format" id="toc-what-is-an-gff-file-general-feature-format" class="nav-link" data-scroll-target="#what-is-an-gff-file-general-feature-format">What is an GFF file (General Feature Format)</a></li>
  <li><a href="#gff-file-analysis" id="toc-gff-file-analysis" class="nav-link" data-scroll-target="#gff-file-analysis">GFF file analysis</a></li>
  <li><a href="#loading-resistance-status-metadata" id="toc-loading-resistance-status-metadata" class="nav-link" data-scroll-target="#loading-resistance-status-metadata">Loading resistance status metadata</a></li>
  <li><a href="#what-is-a-vcf-file-variant-call-format" id="toc-what-is-a-vcf-file-variant-call-format" class="nav-link" data-scroll-target="#what-is-a-vcf-file-variant-call-format">What is a VCF file (Variant Call Format)?</a></li>
  <li><a href="#extracting-gene-region-of-interest-from-vcf-file" id="toc-extracting-gene-region-of-interest-from-vcf-file" class="nav-link" data-scroll-target="#extracting-gene-region-of-interest-from-vcf-file">Extracting gene region of interest from VCF file</a></li>
  <li><a href="#loading-vcf-file-and-calculating-nucleotide-diversity-tajimas-pi" id="toc-loading-vcf-file-and-calculating-nucleotide-diversity-tajimas-pi" class="nav-link" data-scroll-target="#loading-vcf-file-and-calculating-nucleotide-diversity-tajimas-pi">Loading VCF file and calculating nucleotide diversity (Tajimas Pi)</a></li>
  <li><a href="#an-alternative-possibilty-is-to-go-through-constructing-the-genealogies" id="toc-an-alternative-possibilty-is-to-go-through-constructing-the-genealogies" class="nav-link" data-scroll-target="#an-alternative-possibilty-is-to-go-through-constructing-the-genealogies">An alternative possibilty is to go through constructing the genealogies</a></li>
  </ul>
<div class="toc-actions"><div><i class="bi bi-github"></i></div><div class="action-links"><p><a href="https://github.com/kevinkorfmann/workshop-kenya/issues/new" class="toc-action">Report an issue</a></p></div></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title d-none d-lg-block">1. Session - MalariaGEN</h1>
</div>

<div>
  <div class="description">
    The first session about the notebook is about exploring <code>https://www.malariagen.net/</code>.
  </div>
</div>


<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<!-- WARNING: THIS FILE WAS AUTOGENERATED! DO NOT EDIT! -->
<p>The database includes information on the genetic diversity of the parasite, including data on single nucleotide polymorphisms (SNPs) and copy number variations (CNVs). It also provides information on the distribution of drug resistance mutations, which is critical for developing effective malaria treatments.</p>
<p>In addition to genomic data, MalariaGen.net also provides access to other types of data, including epidemiological and clinical data. These data are integrated with the genomic data to provide a comprehensive view of the malaria landscape, including information on the distribution of different malaria parasite strains and their associated clinical outcomes.</p>
<p>Overall, MalariaGen.net is an essential resource for researchers and public health officials working towards the goal of eliminating malaria. By providing open access to comprehensive genomic and epidemiological data, MalariaGen.net is helping to drive advances in our understanding of the biology of the malaria parasite and to inform the development of new and effective malaria control strategies.</p>
<section id="data" class="level2">
<h2 class="anchored" data-anchor-id="data">Data</h2>
<p>https://www.dropbox.com/sh/prelp6g1g369uec/AACruu4uCLf-v0PP1La5Yh_6a?dl=0</p>
</section>
<section id="outline" class="level2">
<h2 class="anchored" data-anchor-id="outline">Outline</h2>
<p>We’ll be focussing on the pathogen P. falciparum.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="./images/laveran_malaria.jpg" class="img-fluid figure-img"></p>
<p></p><figcaption class="figure-caption">Alphonse Laveran was the first to identify the parasite in 1880, and named it Oscillaria malariae.</figcaption><p></p>
</figure>
</div>
<ol type="1">
<li><p>Explore datasets available MalariaGEN related to P. falciparum<br>
We will be looking specifically at the following resources, short variants (SNPs only) and metadata:<br>
- 20,864 samples from 33 countries, but we will filter specifically for african countries (Kenya, …)<br>
- VCF accessable at: ftp://ngs.sanger.ac.uk/production/malaria/Resource/34/Pf7_vcf/ using i.e.&nbsp;FileZilla<br>
- Filter version will be supplied due to large file sizes &gt; 100 GBs (BCFTools) - Resistance status to drugs, like: Chloroquine, Pyrimethamine, Sulfadoxine, Mefloquine, Artemisinin, Piperaquine<br>
- Other metadata, i.e.&nbsp;geographic location</p></li>
<li><p>After downloading and filtering the data we can load them with the python package <code>scikit-allel</code>. Next, we will filter, based on the samples of our country of interest and build geneologies with <code>tsinfer</code>. Afterwards, we date the inferred geneologies and calculate summary statistics, like genetic diversity and Tajimas D.</p></li>
</ol>
<p>To see more available data sources, of humans, mosquitos or different P. vivax feel free to explore: https://www.malariagen.net/data</p>
</section>
<section id="analysis" class="level2">
<h2 class="anchored" data-anchor-id="analysis">Analysis</h2>
<p>In the following part of potential analysis path is shown, feel free to follow it and/or modify it.</p>
<section id="metadata" class="level3">
<h3 class="anchored" data-anchor-id="metadata">Metadata</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co">#pip install requests pandas</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># python package imports</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> os, io, requests</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> pathlib <span class="im">import</span> Path</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a><span class="co">#data_directory = Path("../data")</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Downloading samples information  </span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>samples_url <span class="op">=</span> <span class="st">"https://www.malariagen.net/sites/default/files/Pf7_samples.txt"</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>samples_metadata <span class="op">=</span> pd.read_csv(io.StringIO(requests.get(samples_url).content.decode(<span class="st">'utf-8'</span>)), sep<span class="op">=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>)</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>samples_metadata.head().iloc[:<span class="dv">1</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">

<div>

<table class="dataframe table table-sm table-striped">
  <thead>
    <tr>
      <th></th>
      <th>Sample</th>
      <th>Study</th>
      <th>Country</th>
      <th>Admin level 1</th>
      <th>Country latitude</th>
      <th>Country longitude</th>
      <th>Admin level 1 latitude</th>
      <th>Admin level 1 longitude</th>
      <th>Year</th>
      <th>ENA</th>
      <th>All samples same case</th>
      <th>Population</th>
      <th>% callable</th>
      <th>QC pass</th>
      <th>Exclusion reason</th>
      <th>Sample type</th>
      <th>Sample was in Pf6</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>FP0008-C</td>
      <td>1147-PF-MR-CONWAY</td>
      <td>Mauritania</td>
      <td>Hodh el Gharbi</td>
      <td>20.265149</td>
      <td>-10.337093</td>
      <td>16.565426</td>
      <td>-9.832345</td>
      <td>2014.0</td>
      <td>ERR1081237</td>
      <td>FP0008-C</td>
      <td>AF-W</td>
      <td>82.16</td>
      <td>True</td>
      <td>Analysis_set</td>
      <td>gDNA</td>
      <td>True</td>
    </tr>
  </tbody>
</table>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>african_countries <span class="op">=</span> [<span class="st">'Benin'</span>, <span class="st">'Burkina Faso'</span>,<span class="st">'Cameroon'</span>, <span class="st">'Gabon'</span> ,<span class="st">'Cameroon'</span>, <span class="st">'Côte d</span><span class="ch">\'</span><span class="st">Ivoire'</span>,<span class="st">'Mozambique'</span>, <span class="st">'Democratic Republic of the Congo'</span>,<span class="st">'Sudan'</span>, <span class="st">'Ethiopia'</span>, <span class="st">'Gambia'</span>, <span class="st">'Ghana'</span>, <span class="st">'Guinea'</span>, <span class="st">'Kenya'</span>, <span class="st">'Madagascar'</span>, <span class="st">'Malawi'</span>, <span class="st">'Mali'</span>, <span class="st">'Mauritania'</span>, <span class="st">'Nigeria'</span>, <span class="st">'Senegal'</span>, <span class="st">'Tanzania'</span>, <span class="st">'Uganda'</span>]</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>non_african_countries <span class="op">=</span> [<span class="st">'Bangladesh'</span>, <span class="st">'Cambodia'</span>, <span class="st">'Colombia'</span>, <span class="st">'India'</span>, <span class="st">'Indonesia'</span>, <span class="st">'Laos'</span>, <span class="st">'Myanmar'</span>, <span class="st">'Papua New Guinea'</span>, <span class="st">'Peru'</span>, <span class="st">'Thailand'</span>, <span class="st">'Venezuela'</span>, <span class="st">'Vietnam'</span>]</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>african_subset <span class="op">=</span> [<span class="st">'Kenya'</span>, <span class="st">'Ethiopia'</span>, <span class="st">'Tanzania'</span>, <span class="st">'Uganda'</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Next we identify the sample ids, which we want to explore:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># form sample table above create mask to filter the table to only contain rows of interest</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>samples_metadata_mask <span class="op">=</span> [<span class="va">True</span> <span class="cf">if</span> criteria <span class="kw">in</span> african_subset <span class="cf">else</span> <span class="va">False</span> <span class="cf">for</span> criteria <span class="kw">in</span> samples_metadata.Country]</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>samples_metadata <span class="op">=</span> samples_metadata[samples_metadata_mask]</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a><span class="co"># quality control filtering</span></span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>samples_metadata <span class="op">=</span> samples_metadata[samples_metadata[<span class="st">"QC pass"</span>]]</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a><span class="co"># we only look at genomic DNA (gDNA) for now</span></span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a>samples_metadata_mask <span class="op">=</span> [<span class="va">True</span> <span class="cf">if</span> criteria <span class="kw">in</span> [<span class="st">"gDNA"</span>] <span class="cf">else</span> <span class="va">False</span> <span class="cf">for</span> criteria <span class="kw">in</span> samples_metadata[<span class="st">"Sample type"</span>]]</span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a>samples_metadata <span class="op">=</span> samples_metadata[samples_metadata_mask]</span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a><span class="co"># reseting tables indices</span></span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a>samples_metadata <span class="op">=</span> samples_metadata.reset_index(drop<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>samples_metadata_kenya <span class="op">=</span> samples_metadata[samples_metadata[<span class="st">"Country"</span>] <span class="op">==</span> <span class="st">"Kenya"</span>]</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Number of gDNA samples in Kenya with QC pass True </span><span class="sc">{</span>samples_metadata_kenya<span class="sc">.</span>shape[<span class="dv">0</span>]<span class="sc">}</span><span class="ss">."</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Number of gDNA samples in Kenya with QC pass True 321.</code></pre>
</div>
</div>
<p>Save Kenyan samples ids to txt file:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>kenyian_samples <span class="op">=</span> samples_metadata_kenya[<span class="st">"Sample"</span>].tolist()</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="co">#samples_ids = samples_metadata_kenya.Sample.tolist()</span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="co">#with open(data_directory/"kenyan_sample_ids.txt", "a") as f:</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a><span class="co">#    for samples_id in samples_ids:</span></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a><span class="co">#        f.write(samples_id + "\n")</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="what-is-an-gff-file-general-feature-format" class="level2">
<h2 class="anchored" data-anchor-id="what-is-an-gff-file-general-feature-format">What is an GFF file (General Feature Format)</h2>
<p>Standardized file format, which contains information where to find what genomic element in the genome.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="./images/gff.png" class="img-fluid figure-img"></p>
<p></p><figcaption class="figure-caption">GFF</figcaption><p></p>
</figure>
</div>
</section>
<section id="gff-file-analysis" class="level2">
<h2 class="anchored" data-anchor-id="gff-file-analysis">GFF file analysis</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># path needs to be adapted</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>gff_path <span class="op">=</span> Path(<span class="st">"../../Pfalciparum_replace_Pf3D7_MIT_v3_with_Pf_M76611.gff"</span>)</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>gff <span class="op">=</span> pd.read_csv(gff_path, sep<span class="op">=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, comment<span class="op">=</span><span class="st">"#"</span>, header<span class="op">=</span><span class="va">None</span>)</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a><span class="co"># renaming some columns</span></span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>gff.columns <span class="op">=</span> [<span class="st">"chr"</span>, <span class="dv">1</span>, <span class="st">"region"</span>, <span class="st">"start"</span>, <span class="st">"end"</span>, <span class="dv">4</span>, <span class="st">"strand"</span>, <span class="dv">5</span>, <span class="st">"desc"</span>]</span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a><span class="co"># filtering for gene regions</span></span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a>gff_genes <span class="op">=</span> gff[gff[<span class="st">"region"</span>] <span class="op">==</span> <span class="st">"gene"</span>]</span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a>gff_genes.reset_index(drop<span class="op">=</span><span class="va">True</span>, inplace<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Next we extract genes related to resistance of Malaria drugs.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> re</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>loci_locations <span class="op">=</span> []</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>locis <span class="op">=</span> [<span class="st">"crt"</span>, <span class="st">"dhfr"</span>, <span class="st">"dhps"</span>, <span class="st">"mdr1"</span>, <span class="st">"Kelch13"</span>, <span class="st">"Plasmepsin"</span>]</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> loci <span class="kw">in</span> locis:</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>    pattern <span class="op">=</span> <span class="st">'Name=(.*'</span> <span class="op">+</span> loci.lower() <span class="op">+</span>  <span class="st">'.*|.*'</span> <span class="op">+</span> loci.upper()  <span class="op">+</span> <span class="st">'.*|'</span> <span class="op">+</span> loci <span class="op">+</span> <span class="st">').*;'</span></span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>    mask <span class="op">=</span> [<span class="va">True</span> <span class="cf">if</span> <span class="bu">len</span>(re.findall(pattern, desc)) <span class="op">&gt;</span> <span class="dv">0</span> <span class="cf">else</span> <span class="va">False</span> <span class="cf">for</span> desc <span class="kw">in</span> gff_genes[<span class="st">"desc"</span>]]</span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>    loci_locations.append(gff_genes[mask])</span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a>loci_locations <span class="op">=</span> pd.concat(loci_locations)</span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Plasmepsin not found (mostlikely different name)</span></span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a>loci_locations.reset_index(drop<span class="op">=</span><span class="va">True</span>, inplace<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> _, item <span class="kw">in</span> loci_locations.iterrows():</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(item[<span class="st">"chr"</span>], item[<span class="st">"start"</span>], item[<span class="st">"end"</span>], item[<span class="st">"end"</span>] <span class="op">-</span> item[<span class="st">"start"</span>],  item[<span class="st">"desc"</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Pf3D7_07_v3 402385 406341 3956 ID=PF3D7_0709000;Name=CRT;previous_systematic_id=MAL7P1.27
Pf3D7_04_v3 747897 750065 2168 ID=PF3D7_0417200;Name=DHFR-TS;previous_systematic_id=PFD0830w,MAL4P1.161;synonym=PfDHFR-TS%3Bcurrent%3Dfalse
Pf3D7_08_v3 547896 551057 3161 ID=PF3D7_0810800;Name=PPPK-DHPS;previous_systematic_id=PF08_0095
Pf3D7_05_v3 955955 963095 7140 ID=PF3D7_0523000;Name=MDR1;previous_systematic_id=PFE1150w,MAL5P1.230;synonym=Pgh1,ABCB1
Pf3D7_13_v3 1724600 1727877 3277 ID=PF3D7_1343700;Name=Kelch13;previous_systematic_id=PF13_0238;synonym=K13</code></pre>
</div>
</div>
</section>
<section id="loading-resistance-status-metadata" class="level2">
<h2 class="anchored" data-anchor-id="loading-resistance-status-metadata">Loading resistance status metadata</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>resistance_url <span class="op">=</span> <span class="st">"https://www.malariagen.net/sites/default/files/Pf7_inferred_resistance_status_classification.txt"</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>resistance_status <span class="op">=</span> pd.read_csv(io.StringIO(requests.get(resistance_url).content.decode(<span class="st">'utf-8'</span>)), sep<span class="op">=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>)</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>resistent_samples <span class="op">=</span> []</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>sensitive_samples <span class="op">=</span> []</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>undetermined_samples <span class="op">=</span> []</span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> sample <span class="kw">in</span> kenyian_samples:</span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>    sulfadoxine <span class="op">=</span> resistance_status[resistance_status[<span class="st">"Sample"</span>] <span class="op">==</span> sample][<span class="st">"Sulfadoxine"</span>]</span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> <span class="bu">len</span>(sulfadoxine) <span class="op">==</span> <span class="dv">1</span>:</span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a>        sulfadoxine <span class="op">=</span> sulfadoxine.item()</span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-13"><a href="#cb13-13" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> sulfadoxine <span class="op">==</span> <span class="st">"Resistant"</span>:</span>
<span id="cb13-14"><a href="#cb13-14" aria-hidden="true" tabindex="-1"></a>            resistent_samples.append(sample)</span>
<span id="cb13-15"><a href="#cb13-15" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> sulfadoxine <span class="op">==</span> <span class="st">"Sensitive"</span>:</span>
<span id="cb13-16"><a href="#cb13-16" aria-hidden="true" tabindex="-1"></a>            sensitive_samples.append(sample)</span>
<span id="cb13-17"><a href="#cb13-17" aria-hidden="true" tabindex="-1"></a>        <span class="cf">else</span>:</span>
<span id="cb13-18"><a href="#cb13-18" aria-hidden="true" tabindex="-1"></a>            undetermined_samples.append(sample)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="what-is-a-vcf-file-variant-call-format" class="level2">
<h2 class="anchored" data-anchor-id="what-is-a-vcf-file-variant-call-format">What is a VCF file (Variant Call Format)?</h2>
<p>The Variant Call Format (VCF) specifies the format of a text file used in bioinformatics for storing gene sequence variations. The format has been developed with the advent of large-scale genotyping and DNA sequencing projects, such as the 1000 Genomes Project.</p>
<pre><code>##fileformat=VCFv4.3
##fileDate=20090805
##source=myImputationProgramV3.1
##reference=file:///seq/references/1000GenomesPilot-NCBI36.fasta
##contig=&lt;ID=20,length=62435964,assembly=B36,md5=f126cdf8a6e0c7f379d618ff66beb2da,species="Homo sapiens",taxonomy=x&gt;
##phasing=partial
##INFO=&lt;ID=NS,Number=1,Type=Integer,Description="Number of Samples With Data"&gt;
##INFO=&lt;ID=DP,Number=1,Type=Integer,Description="Total Depth"&gt;
##INFO=&lt;ID=AF,Number=A,Type=Float,Description="Allele Frequency"&gt;
##INFO=&lt;ID=AA,Number=1,Type=String,Description="Ancestral Allele"&gt;
##INFO=&lt;ID=DB,Number=0,Type=Flag,Description="dbSNP membership, build 129"&gt;
##INFO=&lt;ID=H2,Number=0,Type=Flag,Description="HapMap2 membership"&gt;
##FILTER=&lt;ID=q10,Description="Quality below 10"&gt;
##FILTER=&lt;ID=s50,Description="Less than 50% of samples have data"&gt;
##FORMAT=&lt;ID=GT,Number=1,Type=String,Description="Genotype"&gt;
##FORMAT=&lt;ID=GQ,Number=1,Type=Integer,Description="Genotype Quality"&gt;
##FORMAT=&lt;ID=DP,Number=1,Type=Integer,Description="Read Depth"&gt;
##FORMAT=&lt;ID=HQ,Number=2,Type=Integer,Description="Haplotype Quality"&gt;
#CHROM POS      ID         REF   ALT    QUAL  FILTER   INFO                             FORMAT       NA00001         NA00002          NA00003
20     14370    rs6054257  G     A      29    PASS    NS=3;DP=14;AF=0.5;DB;H2           GT:GQ:DP:HQ  0|0:48:1:51,51  1|0:48:8:51,51   1/1:43:5:.,.
20     17330    .          T     A      3     q10     NS=3;DP=11;AF=0.017               GT:GQ:DP:HQ  0|0:49:3:58,50  0|1:3:5:65,3     0/0:41:3
20     1110696  rs6040355  A     G,T    67    PASS    NS=2;DP=10;AF=0.333,0.667;AA=T;DB GT:GQ:DP:HQ  1|2:21:6:23,27  2|1:2:0:18,2     2/2:35:4
20     1230237  .          T     .      47    PASS    NS=3;DP=13;AA=T                   GT:GQ:DP:HQ  0|0:54:7:56,60  0|0:48:4:51,51   0/0:61:2
20     1234567  microsat1  GTC   G,GTCT 50    PASS    NS=3;DP=9;AA=G                    GT:GQ:DP     0/1:35:4        0/2:17:2         1/1:40:3</code></pre>
</section>
<section id="extracting-gene-region-of-interest-from-vcf-file" class="level2">
<h2 class="anchored" data-anchor-id="extracting-gene-region-of-interest-from-vcf-file">Extracting gene region of interest from VCF file</h2>
<p>To create a vcf file which contains only PASS bi-allelic coding SNPs with VQSLOD &gt; 6:</p>
<pre class="console"><code>
bcftools view \
--include 'FILTER="PASS" &amp;&amp; N_ALT=1 &amp;&amp; CDS==1 &amp;&amp; TYPE="snp" &amp;&amp; VQSLOD&gt;6.0' \
--regions Pf3D7_04_v3:600000-1000000 \
--output-type z \
--output-file DHFR-TS.vcf.gz \
Pf3D7_04_v3.pf7.vcf.gz
bcftools index --tbi DHFR-TS.vcf.gz


bcftools view \
--include 'FILTER="PASS" &amp;&amp; N_ALT=1 &amp;&amp; CDS==1 &amp;&amp; TYPE="snp" &amp;&amp; VQSLOD&gt;6.0' \
--regions Pf3D7_05_v3:800000-1100000 \
--output-type z \
--output-file MDR1_surrounding_region.vcf.gz \
Pf3D7_05_v3.pf7.vcf.gz
bcftools index --tbi MDR1_surrounding_region.vcf.gz
</code></pre>
</section>
<section id="loading-vcf-file-and-calculating-nucleotide-diversity-tajimas-pi" class="level2">
<h2 class="anchored" data-anchor-id="loading-vcf-file-and-calculating-nucleotide-diversity-tajimas-pi">Loading VCF file and calculating nucleotide diversity (Tajimas Pi)</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co">#pip install tsinfer tsdate scikit-allel numba matplotlib seaborn</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> tsinfer, tsdate, allel</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> seaborn <span class="im">as</span> sns</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="sourceCode" id="cb18"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co"># change the path to the file accordingly</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a><span class="co"># in windows the file path looks something like 'C:/mydir'</span></span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>vcf_file <span class="op">=</span> <span class="bu">str</span>(<span class="st">"../../datasets/malaria-chromosomes/DHFR-TS_surrounding_region.vcf.gz"</span>)</span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>callset <span class="op">=</span> allel.read_vcf(vcf_file, fields<span class="op">=</span>[<span class="st">'samples'</span>, <span class="st">'calldata/GT'</span>, <span class="st">'variants/ALT'</span>, <span class="st">'variants/CHROM'</span>,</span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a>                                            <span class="st">'variants/FILTER_PASS'</span>, <span class="st">'variants/ID'</span>, <span class="st">'variants/POS'</span>,</span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a>                                            <span class="st">'variants/QUAL'</span>, <span class="st">'variants/REF'</span>,<span class="st">'calldata/GT'</span>,<span class="st">'calldata/PS'</span>])</span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a>gt <span class="op">=</span> allel.GenotypeArray(callset[<span class="st">'calldata/GT'</span>])</span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true" tabindex="-1"></a>phase_sets <span class="op">=</span> callset[<span class="st">'calldata/PS'</span>]</span>
<span id="cb18-12"><a href="#cb18-12" aria-hidden="true" tabindex="-1"></a>pos <span class="op">=</span> callset[<span class="st">"variants/POS"</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> get_genotypes(samples, genotypes):</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>    mask <span class="op">=</span> [<span class="va">True</span> <span class="cf">if</span> sample <span class="kw">in</span> samples <span class="cf">else</span> <span class="va">False</span> <span class="cf">for</span> sample <span class="kw">in</span> callset[<span class="st">"samples"</span>]]</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>    genotypes <span class="op">=</span> genotypes[:,mask]</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>    gt_np <span class="op">=</span> np.array(genotypes.to_haplotypes())</span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> gt_np</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="sourceCode" id="cb20"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>gt_kenyian <span class="op">=</span> get_genotypes(kenyian_samples, gt)</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>gt_resistant_kenyian <span class="op">=</span> get_genotypes(resistent_samples, gt)</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>gt_sensitive_kenyian <span class="op">=</span> get_genotypes(sensitive_samples, gt)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="sourceCode" id="cb21"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>ac <span class="op">=</span> gt_sensitive_kenyian</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>pi, windows, n_bases_sen, counts_sen <span class="op">=</span> allel.windowed_diversity(pos<span class="op">=</span>pos, ac<span class="op">=</span>ac, size<span class="op">=</span><span class="dv">10_000</span>, start<span class="op">=</span>pos[<span class="dv">0</span>])</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>window_start <span class="op">=</span> pd.DataFrame(windows)[<span class="dv">0</span>]</span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a>sns.lineplot(x<span class="op">=</span>window_start, y<span class="op">=</span>pi, label<span class="op">=</span><span class="st">"Nucleotide Diversity (sensitive)"</span>)</span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a>ac <span class="op">=</span> gt_resistant_kenyian</span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a>pi, windows, n_bases_res, counts_res <span class="op">=</span> allel.windowed_diversity(pos<span class="op">=</span>pos, ac<span class="op">=</span>ac, size<span class="op">=</span><span class="dv">10_000</span>, start<span class="op">=</span>pos[<span class="dv">0</span>])</span>
<span id="cb21-9"><a href="#cb21-9" aria-hidden="true" tabindex="-1"></a>window_start <span class="op">=</span> pd.DataFrame(windows)[<span class="dv">0</span>]</span>
<span id="cb21-10"><a href="#cb21-10" aria-hidden="true" tabindex="-1"></a>sns.lineplot(x<span class="op">=</span>window_start, y<span class="op">=</span>pi, label<span class="op">=</span><span class="st">"Nucleotide Diversity (resistant)"</span>)</span>
<span id="cb21-11"><a href="#cb21-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-12"><a href="#cb21-12" aria-hidden="true" tabindex="-1"></a>plt.axvline(<span class="dv">747897</span>, ls<span class="op">=</span><span class="st">'--'</span>)</span>
<span id="cb21-13"><a href="#cb21-13" aria-hidden="true" tabindex="-1"></a>plt.axvline(<span class="dv">750065</span>, ls<span class="op">=</span><span class="st">'--'</span>)</span>
<span id="cb21-14"><a href="#cb21-14" aria-hidden="true" tabindex="-1"></a>plt.grid(<span class="va">True</span>)</span>
<span id="cb21-15"><a href="#cb21-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-16"><a href="#cb21-16" aria-hidden="true" tabindex="-1"></a>plt.suptitle(<span class="st">"DHFR (Sulfadoxine target)"</span>)</span>
<span id="cb21-17"><a href="#cb21-17" aria-hidden="true" tabindex="-1"></a><span class="co">#plt.savefig("./images/dhfr_scikit.png")</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="./images/dhfr_scikit.png" class="img-fluid figure-img"></p>
<p></p><figcaption class="figure-caption">DHFR gene</figcaption><p></p>
</figure>
</div>
</section>
<section id="an-alternative-possibilty-is-to-go-through-constructing-the-genealogies" class="level2">
<h2 class="anchored" data-anchor-id="an-alternative-possibilty-is-to-go-through-constructing-the-genealogies">An alternative possibilty is to go through constructing the genealogies</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> get_tree_sequence_from_genotypes(samples, genotypes):</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>    mask <span class="op">=</span> [<span class="va">True</span> <span class="cf">if</span> sample <span class="kw">in</span> samples <span class="cf">else</span> <span class="va">False</span> <span class="cf">for</span> sample <span class="kw">in</span> callset[<span class="st">"samples"</span>]]</span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>    genotypes <span class="op">=</span> genotypes[:,mask]</span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a>    gt_np <span class="op">=</span> np.array(genotypes.to_haplotypes())</span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-7"><a href="#cb22-7" aria-hidden="true" tabindex="-1"></a>    new_pos <span class="op">=</span> pos <span class="op">-</span> pos[<span class="dv">0</span>] </span>
<span id="cb22-8"><a href="#cb22-8" aria-hidden="true" tabindex="-1"></a>    num_sites <span class="op">=</span> gt_np.shape[<span class="dv">0</span>]</span>
<span id="cb22-9"><a href="#cb22-9" aria-hidden="true" tabindex="-1"></a>    <span class="cf">with</span> tsinfer.SampleData(sequence_length<span class="op">=</span>new_pos[<span class="op">-</span><span class="dv">1</span>]<span class="op">+</span><span class="dv">1</span>) <span class="im">as</span> sample_data:</span>
<span id="cb22-10"><a href="#cb22-10" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> i, site <span class="kw">in</span> <span class="bu">enumerate</span>(gt_np):</span>
<span id="cb22-11"><a href="#cb22-11" aria-hidden="true" tabindex="-1"></a>            sample_data.add_site(new_pos[i], site, ancestral_allele<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb22-12"><a href="#cb22-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-13"><a href="#cb22-13" aria-hidden="true" tabindex="-1"></a>    inferred_ts <span class="op">=</span> tsinfer.infer(sample_data)</span>
<span id="cb22-14"><a href="#cb22-14" aria-hidden="true" tabindex="-1"></a>    <span class="co">#inferred_ts_reduced = reduce_tree_sequence(inferred_ts, 500)</span></span>
<span id="cb22-15"><a href="#cb22-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-16"><a href="#cb22-16" aria-hidden="true" tabindex="-1"></a>    inferred_ts <span class="op">=</span> inferred_ts.simplify(keep_unary<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb22-17"><a href="#cb22-17" aria-hidden="true" tabindex="-1"></a>    inferred_dated_ts<span class="op">=</span> tsdate.date(inferred_ts, Ne<span class="op">=</span><span class="dv">2000</span>, mutation_rate<span class="op">=</span><span class="dv">2</span> <span class="op">*</span> <span class="fl">10e-4</span>)</span>
<span id="cb22-18"><a href="#cb22-18" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> inferred_dated_ts</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="sourceCode" id="cb23"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>ts_all <span class="op">=</span> get_tree_sequence_from_genotypes(kenyian_samples, gt)</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>ts_resistant <span class="op">=</span> get_tree_sequence_from_genotypes(resistent_samples, gt)</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>ts_sensitive <span class="op">=</span> get_tree_sequence_from_genotypes(sensitive_samples, gt)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> get_stats(ts, wins):</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>    tajima <span class="op">=</span> ts.Tajimas_D(windows<span class="op">=</span>wins, mode<span class="op">=</span><span class="st">"site"</span>)</span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>    diversity <span class="op">=</span> ts.diversity(windows<span class="op">=</span>wins, mode<span class="op">=</span><span class="st">"site"</span>)</span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> diversity, tajima</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="sourceCode" id="cb25"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>L <span class="op">=</span> ts_all.sequence_length</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>wins <span class="op">=</span> np.linspace(<span class="dv">0</span>, L, <span class="dv">31</span>)</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>mids <span class="op">=</span> (wins[<span class="dv">1</span>:] <span class="op">+</span> wins[:<span class="op">-</span><span class="dv">1</span>]) <span class="op">/</span> <span class="dv">2</span></span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>diversity, tajima <span class="op">=</span> get_stats(ts_all, wins)</span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a>diversity_res, tajima_res <span class="op">=</span> get_stats(ts_resistant, wins)</span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a>diversity_sens, tajima_sens <span class="op">=</span> get_stats(ts_sensitive, wins)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="sourceCode" id="cb26"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>fig, axs <span class="op">=</span> plt.subplots(<span class="dv">2</span>, <span class="dv">2</span>, figsize<span class="op">=</span>(<span class="dv">8</span>, <span class="dv">6</span>))</span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> axs[<span class="dv">0</span>][<span class="dv">0</span>]</span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>sns.lineplot(x<span class="op">=</span>mids, y<span class="op">=</span>diversity_sens, ax<span class="op">=</span>ax, label<span class="op">=</span><span class="st">"Nucleotide diversity"</span>)</span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a>ax.axvline(<span class="dv">747897</span><span class="op">-</span>pos[<span class="dv">0</span>], ls<span class="op">=</span><span class="st">'--'</span>)</span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a>ax.axvline(<span class="dv">750065</span><span class="op">-</span>pos[<span class="dv">0</span>], ls<span class="op">=</span><span class="st">'--'</span>)</span>
<span id="cb26-6"><a href="#cb26-6" aria-hidden="true" tabindex="-1"></a>ax.grid(<span class="va">True</span>)</span>
<span id="cb26-7"><a href="#cb26-7" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> axs[<span class="dv">0</span>][<span class="dv">1</span>]</span>
<span id="cb26-8"><a href="#cb26-8" aria-hidden="true" tabindex="-1"></a>sns.lineplot(x<span class="op">=</span>mids, y<span class="op">=</span>tajima_sens, ax<span class="op">=</span>ax, label<span class="op">=</span><span class="st">"Tajimas D"</span>)</span>
<span id="cb26-9"><a href="#cb26-9" aria-hidden="true" tabindex="-1"></a>ax.set_ylim(<span class="op">-</span><span class="dv">2</span>, <span class="dv">2</span>)</span>
<span id="cb26-10"><a href="#cb26-10" aria-hidden="true" tabindex="-1"></a>ax.axhline(<span class="dv">0</span>)</span>
<span id="cb26-11"><a href="#cb26-11" aria-hidden="true" tabindex="-1"></a>ax.axvline(<span class="dv">747897</span><span class="op">-</span>pos[<span class="dv">0</span>], ls<span class="op">=</span><span class="st">'--'</span>)</span>
<span id="cb26-12"><a href="#cb26-12" aria-hidden="true" tabindex="-1"></a>ax.axvline(<span class="dv">750065</span><span class="op">-</span>pos[<span class="dv">0</span>], ls<span class="op">=</span><span class="st">'--'</span>)</span>
<span id="cb26-13"><a href="#cb26-13" aria-hidden="true" tabindex="-1"></a>ax.grid(<span class="va">True</span>)</span>
<span id="cb26-14"><a href="#cb26-14" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> axs[<span class="dv">1</span>][<span class="dv">0</span>]</span>
<span id="cb26-15"><a href="#cb26-15" aria-hidden="true" tabindex="-1"></a>sns.lineplot(x<span class="op">=</span>mids, y<span class="op">=</span>diversity_res, ax<span class="op">=</span>ax, label<span class="op">=</span><span class="st">"Nucleotide diversity"</span>)</span>
<span id="cb26-16"><a href="#cb26-16" aria-hidden="true" tabindex="-1"></a>ax.axvline(<span class="dv">747897</span><span class="op">-</span>pos[<span class="dv">0</span>], ls<span class="op">=</span><span class="st">'--'</span>)</span>
<span id="cb26-17"><a href="#cb26-17" aria-hidden="true" tabindex="-1"></a>ax.axvline(<span class="dv">750065</span><span class="op">-</span>pos[<span class="dv">0</span>], ls<span class="op">=</span><span class="st">'--'</span>)</span>
<span id="cb26-18"><a href="#cb26-18" aria-hidden="true" tabindex="-1"></a>ax.grid(<span class="va">True</span>)</span>
<span id="cb26-19"><a href="#cb26-19" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> axs[<span class="dv">1</span>][<span class="dv">1</span>]</span>
<span id="cb26-20"><a href="#cb26-20" aria-hidden="true" tabindex="-1"></a>sns.lineplot(x<span class="op">=</span>mids, y<span class="op">=</span>tajima_res, ax<span class="op">=</span>ax, label<span class="op">=</span><span class="st">"Tajimas D"</span>)</span>
<span id="cb26-21"><a href="#cb26-21" aria-hidden="true" tabindex="-1"></a>ax.set_ylim(<span class="op">-</span><span class="dv">2</span>, <span class="dv">2</span>)</span>
<span id="cb26-22"><a href="#cb26-22" aria-hidden="true" tabindex="-1"></a>ax.axhline(<span class="dv">0</span>)</span>
<span id="cb26-23"><a href="#cb26-23" aria-hidden="true" tabindex="-1"></a>ax.axvline(<span class="dv">747897</span><span class="op">-</span>pos[<span class="dv">0</span>], ls<span class="op">=</span><span class="st">'--'</span>)</span>
<span id="cb26-24"><a href="#cb26-24" aria-hidden="true" tabindex="-1"></a>ax.axvline(<span class="dv">750065</span><span class="op">-</span>pos[<span class="dv">0</span>], ls<span class="op">=</span><span class="st">'--'</span>)</span>
<span id="cb26-25"><a href="#cb26-25" aria-hidden="true" tabindex="-1"></a>ax.grid(<span class="va">True</span>)</span>
<span id="cb26-26"><a href="#cb26-26" aria-hidden="true" tabindex="-1"></a>fig.suptitle(<span class="st">"DHFR (Sulfadoxine target - sensitive [top] resistant [bottom])"</span>)</span>
<span id="cb26-27"><a href="#cb26-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-28"><a href="#cb26-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-29"><a href="#cb26-29" aria-hidden="true" tabindex="-1"></a><span class="co">#plt.savefig("./images/dhfr.png")</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="./images/dhfr.png" class="img-fluid figure-img"></p>
<p></p><figcaption class="figure-caption">DHFR gene</figcaption><p></p>
</figure>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>